core/src/test/java/nl/nn/adapterframework/pipes/PgpPipeTest.java

```diff
@@ -0,0 +1,203 @@
+						new String[]{"decrypt", recipient[2], "wrong password :/", null, null, null}},
+				{"Decrypt Plaintext", "org.bouncycastle.openpgp.PGPException",
+						new String[]{"decrypt", recipient[2], recipient[1], null, null, null},
+						new String[]{"decrypt", recipient[2], recipient[1], null, null, null}},
+		});
+	}
+
+	public PgpPipeTest(String name, String expectation, String[] encryptParams, String[] decryptParams) {
+		setup();
+		this.expectation = expectation;
+		this.encryptParams = encryptParams;
+		this.decryptParams = decryptParams;
+	}
+
+	@Test
+	public void dotest() throws Throwable {
+		try {
+			// Configure pipes
+			configurePipe(encryptPipe, encryptParams);
+			configurePipe(decryptPipe, decryptParams);
+
+			// Encryption phase
+			PipeRunResult encryptionResult = encryptPipe.doPipe(MESSAGE, session);
+
+			// Make sure it's PGP message
+			OutputStream mid = (OutputStream) encryptionResult.getResult();
+			assertMessage(mid.toString(), MESSAGE);
+
+			// Decryption phase
+			PipeRunResult decryptionResult = decryptPipe.doPipe(encryptionResult.getResult(), session);
+			OutputStream result = (OutputStream) decryptionResult.getResult();
+
+			// Assert decrypted message equals to the original message
+			Assert.assertEquals(MESSAGE, result.toString());
+			Assert.assertEquals("success", expectation);
+		} catch (Exception e) {
+			if (checkExceptionClass(e, expectation)) {
+				Assert.assertTrue(true);
+			} else {
+				throw e;
+			}
+		}
+	}
+
+	/**
+	 * Creates pipes and pipeline session base for testing.
+	 */
+	public void setup() {
+		session = new PipeLineSessionBase();
+
+		encryptPipe = new PGPPipe();
+		decryptPipe = new PGPPipe();
+
+		encryptPipe.registerForward(new PipeForward("success", null));
+		encryptPipe.setName(encryptPipe.getClass().getSimpleName() + " under test");
+
+		decryptPipe.registerForward(new PipeForward("success", null));
+		decryptPipe.setName(decryptPipe.getClass().getSimpleName() + " under test");
+	}
+
+	/**
+	 * Sets the parameters of the pipes
+	 * @param pipe Pipe to be configured.
+	 * @param params Parameters to be set.
+	 * @throws ConfigurationException When there's an exception during configuration.
+	 */
+	private void configurePipe(PGPPipe pipe, String[] params) throws ConfigurationException {
+		// Just so we dont have to change numbers every time we change order.
+		int i = 0;
+		pipe.setAction(params[i++]);
+		pipe.setSecretKey(addFolderPath(params[i++]));
+		pipe.setSecretPassword(params[i++]);
+		pipe.setPublicKey(addFolderPath(params[i++]));
+		pipe.setSenders(params[i++]);
+		pipe.setRecipients(params[i]);
+		pipe.configure();
+	}
+
+	/**
+	 * Adds folder's path to every file in the given parameters.
+	 * @param param A list of files separated by semicolon.
+	 * @return A list of files separated by semicolon including the parent folder's path.
+	 */
+	private String addFolderPath(String param) {
+		if (param == null)
+			return null;
+		String[] keys = param.split(";");
```

Comment:
You can also use a tokenizer, which trims the results, and gives you a iterator to directly iterate over :)

Reply:
Yep, and it also doesn't require me to hold all the keys in memory, but when I use it I can't set stringbuilder's capacity, which decreases performance since it needs to scale. So memory vs time :P . But with tokenizer it looks cleaner so I changed it.